### PR TITLE
[Parallel Executor] MVHashMap refactoring to separate data / modules

### PR DESCRIPTION
**Pull request by gelash**
Separate out data & modules - useful for executable / versioned code work, but also for caching MoveValues / CompiledModules, as these types (and handling) might be different.

Introduced some Executable tooling, but a separate diff will incorporate module versioning and executable treatment in parallel executor (e.g. validation based on fetching, view traits) & respective testing.

### CODE REVIEW

File: types/src/executable.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{access_path::AccessPath, state_store::state_key::StateKey};
+use aptos_crypto::HashValue;
+use std::sync::Arc;
+
+#[derive(PartialEq)]
+pub enum ExecutableDescriptor {
+    /// Possibly speculative, based on code published during the block.
+    Block(HashValue),
```
Comment (georgemitenkov) [inline types/src/executable.rs:11]: `Published(HashValue)`?

Comment (gelash) [reply types/src/executable.rs:11]: maybe `Publishing(HashValue)` would be even better, wdyt?

File: aptos-move/mvhashmap/src/lib.rs [hunk 2]
```diff
@@ -1,304 +1,120 @@
...
-                            // Resolve to the write if no deltas were applied in between.
-                            let write_version = (*idx, *incarnation);
-                            return Ok(Version(write_version, data.clone()));
-                        },
-                        (EntryCell::Write(incarnation, data), Some(accumulator)) => {
-                            // Deltas were applied. We must deserialize the value
-                            // of the write and apply the aggregated delta accumulator.
+    /// Read data at access path 'key', from the perspective of transaction 'txn_idx'.
+    pub fn read(&self, key: &K, txn_idx: TxnIndex) -> anyhow::Result<MVDataOutput<V>, MVDataError> {
+        self.data.read(key, txn_idx)
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:92]: I'll change to read_data, good point.
Unbelievable, I was deliberating between dispatching in read for now or using this placeholder-like fetch_module, but somehow I didn't include either? Why does the forge pass? Oh because we never need to read published modules and it comes from storage... LoL I'd have broken everything. Let me dispatch on view's read for now, and add a test that makes sure fetch module returns the correct blob.

Later fetch module will either return blob or executable (if already recorded), and record for validation only when it returns the executable - so VM will store it once compiled - except init_module which actually should just work because it's staged inside the VM's data_store (just need to add Option<Executable> there as well).

[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:92]: oh, LOL, the read write fallback saves it. If a module got published, and we try to read it we anyway fallback to sequential right now. If it didn't get published, the read will not find it which is a correct behavior and will come from storage = )

I should still probably fix it, right? :D but it's hard to write a PE test, can just test on MVHashMap side

[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:92]: @zekun000 actually maybe this is a good plan for rolling out:
- we keep the module read-write fallback even after adding the new fetch & validation flow for the modules. This way, the current VM will keep using the get_state_view to get the module data, and even while landing the new code to aptos-core, things will work as normal.
- we then make changes to Move-VM to start using the new APIs and manage modules/structs/functions/links a bit different, and this is the change that we will test, so it will basically be a versioning of VM new vs old. Once we are confident in new we switch to it fully.
- somewhere here we may want to introduce changes to aptos-vm for testing the new version across the blocks - hopefully just a simple way to know if we are using the new flow, do a few more things it can support to test properly
- finally, we can get rid of the fallback logic from the parallel executor, get rid of the get_state_view interface (have dedicated ones for data / modules), and get rid of all the flushing.

@runtian-zhou @vgao1996 as well.

[the remark this answers is not in the document]
Comment (zekun000) [reply aptos-move/mvhashmap/src/lib.rs:92]: I think keep the read-write fallback is fine, but we probably still want to keep expected behavior on interfaces i.e. read versioned data instead of going to storage and rely on fallbacks?

[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:92]: makes sense - there will be totally new interfaces for modules, but I'll patch the old common interface (for data and code)`get_state_view` to dispatch and return the right versioned blob if it exists. and it may exist because the r/w intersection is checked right after.
maybe we can even use it to assert somehow but not super sure, will look and update the PR today.

File: aptos-move/mvhashmap/src/types.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_aggregator::delta_change_set::DeltaOp;
+use aptos_crypto::hash::HashValue;
+use std::sync::Arc;
+
+pub type TxnIndex = usize;
```
Comment (gelash) [inline aptos-move/mvhashmap/src/types.rs:8]: change this to u32

File: types/src/executable.rs (hunk 1 above)
Comment (gelash) [reply types/src/executable.rs:11]: Actually Published is better, done :-)

File: aptos-move/mvhashmap/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:89]: added things to view.rs, ugly, but temporary

File: aptos-move/block-executor/src/scheduler.rs [hunk 4]
```diff
@@ -164,14 +162,14 @@ impl ValidationStatus {
 
 pub struct Scheduler {
     /// Number of txns to execute, immutable.
-    num_txns: usize,
+    num_txns: TxnIndex,
```
Comment (runtian-zhou) [inline aptos-move/block-executor/src/scheduler.rs:165]: Are these changes needed?

File: aptos-move/mvhashmap/src/types.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_aggregator::delta_change_set::DeltaOp;
```
Comment (runtian-zhou) [inline aptos-move/mvhashmap/src/types.rs:4]: I feel like you should move the executable trait definition to here rather than in `aptos-types`?

File: aptos-move/mvhashmap/src/types.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    Dependency(TxnIndex),
+}
+
+/// Returned as Ok(..) when read successfully from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVDataOutput<V> {
+    /// Result of resolved delta op, always u128. Unlike with `Version`, we return
+    /// actual data because u128 is cheap to copy and validation can be done correctly
+    /// on values as well (ABA is not a problem).
+    Resolved(u128),
```
Comment (runtian-zhou) [inline aptos-move/mvhashmap/src/types.rs:39]: IIUC `resolved` should only be related to aggregator? Just want to double check my understanding here.

File: aptos-move/mvhashmap/src/types.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    /// and not the data to avoid copying big values around.
+    Versioned(Version, Arc<V>),
+}
+
+/// Returned as Ok(..) when read successfully from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVCodeOutput<M, X> {
+    /// Arc to the executable corresponding to the latest module, and an indicator
+    /// whether the executable corresponds to a storage-version module.
+    Executable((Arc<X>, bool)),
```
Comment (runtian-zhou) [inline aptos-move/mvhashmap/src/types.rs:50]: I feel like named struct is better suited here. Sth like
```
Executable {
    binary: Arc<X>,
    is_storage: bool,
}
```

The `is_storage` is probably a bad name but can probably provide ppl with better context.

File: aptos-move/mvhashmap/src/types.rs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    /// Arc to the latest module, together with its (cryptographic) hash. Note that
+    /// this can't be a storage-level module, as it's from multi-versioned code map.
+    Module((Arc<M>, Option<HashValue>)),
```
Comment (runtian-zhou) [inline aptos-move/mvhashmap/src/types.rs:53]: Also I don't quite understand why are we distinguishing Executable and Module? Is the idea that one is for CompiledModule and the other is for a more resolved form? If so I feel like this is a implementation detail that needs to be filled out in move side in your future refactor and not sth this mvhashmap should care about?

File: types/src/executable.rs [hunk 9]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{access_path::AccessPath, state_store::state_key::StateKey};
+use aptos_crypto::HashValue;
+use std::sync::Arc;
+
+#[derive(PartialEq)]
+pub enum ExecutableDescriptor {
```
Comment (runtian-zhou) [inline types/src/executable.rs:9]: See my previous comment. Should we move these types to mvhashmap for now?

**Review by runtian-zhou — COMMENTED** [review]
My biggest question is how much implementation detail we want to put inside mvhashmap's code cache. It does seem to include lots of the VM specific logic, i.e: distinguishing the resolved (Executable) and unresolved format (TransactionWrite). Can we let VM do the job for us?

File: aptos-move/block-executor/src/scheduler.rs (hunk 4 above)
Comment (gelash) [reply aptos-move/block-executor/src/scheduler.rs:165]: oh, these, we noticed that MVHashMap and executor were aliasing the same types, but not even using them everywhere, was a mess. Zekun asked @danielxiangzl to change TxnIndex and Incarnation to u32 in the rolling commit diff and that's when it become apparent.
So while refactoring the MVHashMap also took care of this annoyance. Unfortunately, led to bunch of usize casts because Rust doesn't want to use anything else as an index, but other than that, that's the story.
I can go and remove the changes but would rather not - first it takes latency of weeks to get any change in, second it's mostly mechanical changes that should be easy to skim though - sorry for the trouble though.

File: aptos-move/mvhashmap/src/types.rs (hunk 5 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/types.rs:4]: I have added in types because putting it in aptos-move screams "circular-dependency", as this is a trait that Move-VM has to implement. As you said, maybe we move it to Move-VM in a few PRs, but in the interim this seemed a more logical place due to that reasoning.

File: aptos-move/mvhashmap/src/types.rs (hunk 6 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/types.rs:39]: Yes, resolved is an aggregator concept, that's why it only exists in the data-related part now

File: aptos-move/mvhashmap/src/types.rs (hunk 7 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/types.rs:50]: Fair point, will do

File: aptos-move/mvhashmap/src/types.rs (hunk 8 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/types.rs:53]: CompiledModule / blob are the same thing at this point - we will and should optimize soon to save serialization etc overhead, but afaik that's the only difference.
Executable corresponds to more resolved form, something that will live in a sort of "stable" memory, be linkable, executable, so in the most simple (because our representations currently are pretty basic) I think corresponds to the structs and functions that are currently indexed and linked that way. The first version of the executable can just hold those Arc's directly, I'm thinking - later we will optimize memory consumption & re-use, representation, etc.

Why does MVHashMap need to know - it only knows via the trait, and the trait now mainly exposes the size() -> this is so we can implement a basic memory accounting model to RLU or flush or something. Storage-version executables will be re-used across the blocks, things that got overwritten will be pruned out, and at some point we will also have eviction based on metadata we can track and sizes.

Currently, the simplest thing is to make VersionedCode hold these Executables, and manage their correspondence with the blob (CompiledModules) - it's actually very convenient, because Move-VM will have to generate the executables, but due to concurrency, if you read a blob/CompiledModule, you should not have a dependency on it or need to validate, only once you obtain an executable, because that's the code you'll end up running. There will be "store_executable" function in the trait so when Move-VM doesn't get the executable (because it doesn't exist), it will create an executable, and call store. Then it will read again, until it gets an executable and can proceed.

Then there is a detail of init_module but that works out because data_store does the staging there - those executables won't be stored in MVHashMap via trait, which avoids all the today's problems - these will be just within the context of the transaction and if txn is successful, then the output is propagated, and if we want as an optimization we can also pass the executables there. Anyway, this last part is minor detail.

File: types/src/executable.rs (hunk 9 above)
Comment (gelash) [reply types/src/executable.rs:9]: same reasoning, but in any case this will have to evolve soon.

File: types/src/executable.rs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    // TODO: compiled module when available to avoid deserialization.
+    Executable(Arc<X>),
+}
+
+/// View for the VM for interacting with the multi-versioned executable cache.
+pub trait ExecutableView {
+    type Key;
+    type Executable: Executable;
+
+    /// This is an optimization to bypass transactional semantics and share the
```
Comment (sasha8) [inline types/src/executable.rs:60]: Can you elaborate more on the plan here? What are we bypassing? we will call MVHasmMap anyway, right? Are we going to have a separate data view and code view that both will call MVHashMap?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+/// A VersionedValue internally contains a BTreeMap from indices of transactions
+/// that update the given access path alongside the corresponding entries.
+struct VersionedValue<V> {
+    versioned_map: BTreeMap<TxnIndex, CachePadded<Entry<V>>>,
+
+    // Note: this can cache base (storage) value in Option<u128> to facilitate
+    // aggregator validation & reading in the future, if needed.
+    contains_delta: bool,
```
Comment (sasha8) [inline aptos-move/mvhashmap/src/versioned_data.rs:56]: Unrelated to this diff, but remind me why we need this bool if we have this information in the delta_keys vector?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+const FLAG_DONE: usize = 0;
+const FLAG_ESTIMATE: usize = 1;
+
+/// Every entry in shared multi-version data-structure has an "estimate" flag
+/// and some content.
+struct Entry<V> {
+    /// Used to mark the entry as a "write estimate". Even though the entry
+    /// lives inside the DashMap and the entry access will have barriers, we
+    /// still make the flag Atomic to provide acq/rel semantics on its own.
+    flag: AtomicUsize,
```
Comment (sasha8) [inline aptos-move/mvhashmap/src/versioned_data.rs:31]: rename to "estimation"?

File: aptos-move/mvhashmap/src/versioned_code.rs [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+struct Entry<V: TransactionWrite> {
+    /// Used to mark the entry as a "write estimate". Even though the entry
+    /// lives inside the DashMap and the entry access will have barriers, we
+    /// still make the flag Atomic to provide acq/rel semantics on its own.
+    flag: AtomicUsize,
+
+    /// The contents of the module as produced by the VM (can be WriteOp based on a
+    /// blob or CompiledModule, but must satisfy TransactionWrite to be able to
+    /// generate the hash below.
+    module: Arc<V>,
```
Comment (sasha8) [inline aptos-move/mvhashmap/src/versioned_code.rs:35]: We do not store the executable here?
Is it because the executable is a function of other modules?

Also, if during validation hash is the same but the executable is different because dependencies changed? 
Will I abort when fetching dependencies? What if I do not need them? Then we are fine?

File: aptos-move/mvhashmap/src/versioned_code.rs [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+/// A VersionedValue internally contains a BTreeMap from indices of transactions
+/// that update the given access path alongside the corresponding entries.
+struct VersionedValue<V: TransactionWrite, X: Executable> {
+    versioned_map: BTreeMap<TxnIndex, CachePadded<Entry<V>>>,
+
+    /// Executable based on the storage version of the module.
+    base_executable: Option<Arc<X>>,
+    /// Executables corresponding to published versions of the module, based on hash.
+    executables: HashMap<HashValue, Arc<X>>,
```
Comment (sasha8) [inline aptos-move/mvhashmap/src/versioned_code.rs:50]: Why this is not part of the Entry?
Is this an optimization in case two transactions compute the same executable?
Or you only lazily compute it on a fetch?

File: aptos-move/mvhashmap/src/versioned_code.rs [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+            .mark_estimate();
+    }
+
+    pub(crate) fn delete(&self, key: &K, txn_idx: TxnIndex) {
+        // TODO: investigate logical deletion.
+        let mut v = self.values.get_mut(key).expect("Path must exist");
+        v.versioned_map.remove(&txn_idx);
+    }
+
+    fn read(
```
Comment (sasha8) [inline aptos-move/mvhashmap/src/versioned_code.rs:119]: "read_moudle" to be consistent?

File: aptos-move/mvhashmap/src/versioned_code.rs [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+                })
+            },
+            Err(NotFound) => {
+                match self.values.get(key) {
+                    Some(v) => v
+                        .base_executable
+                        .as_ref()
+                        .map(|x| Executable((x.clone(), true))) // storage-version
+                        .ok_or(NotFound),
+                    None => Err(NotFound),
```
Comment (sasha8) [inline aptos-move/mvhashmap/src/versioned_code.rs:202]: When do you store base_executable? Should not you read from storage now and store? Or the wrapper does it and write back the base executable?

**Review by sasha8 — COMMENTED** [review]
Some comments, mostly questions for my understanding.

File: aptos-move/mvhashmap/src/types.rs [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+/// Returned as Ok(..) when read successfully from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVCodeOutput<M, X> {
+    /// Arc to the executable corresponding to the latest module, and an indicator
+    /// whether the executable corresponds to a storage-version module.
+    Executable((Arc<X>, bool)),
+    /// Arc to the latest module, together with its (cryptographic) hash. Note that
+    /// this can't be a storage-level module, as it's from multi-versioned code map.
```
Comment (sasha8) [inline aptos-move/mvhashmap/src/types.rs:52]: old comment

**Review by sasha8 — APPROVED** [review]
Looking forward to the follow up diifs

File: aptos-move/mvhashmap/src/types.rs (hunk 7 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/types.rs:50]: bool had to be ExecutableDescriptor, which I think makes things cleaner. 
This output has to be adjusted for VM trait anyway, so I don't want to introduce too many intermediate types there. Also, we will likely adjust them as we fill in the implementations in the next PRs.

File: types/src/executable.rs (hunk 10 above)
Comment (gelash) [reply types/src/executable.rs:60]: When we produce an executable based on some module (that isn't from the running transaction), we let MVHashMap know immediately instead of waiting for the transaction to finish.

File: aptos-move/mvhashmap/src/versioned_data.rs (hunk 11 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_data.rs:56]: it's currently just saving the unnecessary locking on delta_keys, and allows that to be a Vec. not too important, but also relevant meta-data that we can store in an entry so why not.

File: aptos-move/mvhashmap/src/versioned_data.rs (hunk 12 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_data.rs:31]: it had _DONE and _ESTIMATE values so can't just rename.
can make an atomicboolean and then it can be "estimation_flag" or smt. Do you prefer that?

File: aptos-move/mvhashmap/src/versioned_code.rs (hunk 13 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_code.rs:35]: we don't care about dependencies for now.
executable is a function of module, different version of VM may produce different executable based on the same module (code). module is code.

File: aptos-move/mvhashmap/src/versioned_code.rs (hunk 14 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_code.rs:50]: It is so we don't recompute the executable every time we need it (based on the same code), yes

Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_code.rs:50]: it's basically a cache of executables based on the module (code) cryptographic hash, or for the storage version (can't change)

File: aptos-move/mvhashmap/src/versioned_code.rs (hunk 16 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_code.rs:202]: the VM is the only one that can produce executables, so storage or not it will go through the "wrapper" and store the executable. Storage version doesn't change so no need for the hash as key.

File: aptos-move/mvhashmap/src/versioned_code.rs (hunk 15 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_code.rs:119]: not sure I follow anymore, it's a private fn.
consistent with what?

File: aptos-move/mvhashmap/src/types.rs (hunk 3 above)
Comment (sgilflores3) [reply aptos-move/mvhashmap/src/types.rs]: K thanks

File: aptos-move/mvhashmap/src/versioned_code.rs (hunk 15 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_code.rs]: read_data and read_code now

File: aptos-move/block-executor/src/proptest_types/tests.rs [hunk 18]
```diff
@@ -406,7 +406,7 @@ fn publishing_fixed_params() {
 // Test a single transaction intersection interleaves with a lot of dependencies and
 // not overlapping module r/w keys.
 fn module_publishing_races() {
-    for _ in 0..10 {
+    for _ in 0..5 {
```
Comment (zekun000) [inline aptos-move/block-executor/src/proptest_types/tests.rs:409]: why?

File: aptos-move/block-executor/src/scheduler.rs [hunk 19]
```diff
@@ -165,14 +163,14 @@ impl ValidationStatus {
...
 
     /// A shared index that tracks the minimum of all transaction indices that require execution.
     /// The threads increment the index and attempt to create an execution task for the corresponding
     /// transaction, if the status of the txn is 'ReadyToExecute'. This implements a counting-based
     /// concurrent ordered set. It is reduced as necessary when transactions become ready to be
     /// executed, in particular, when execution finishes and dependencies are resolved.
-    execution_idx: AtomicUsize,
+    execution_idx: AtomicU32,
```
Comment (zekun000) [inline aptos-move/block-executor/src/scheduler.rs:173]: the atomic discussion reminds me, do we need to pad the two idx too?

File: aptos-move/mvhashmap/src/lib.rs [hunk 20]
```diff
@@ -2,304 +2,126 @@
...
+    // Functions shared for data and code.
 
-impl<K: Hash + Clone + Eq, V: TransactionWrite> MVHashMap<K, V> {
-    pub fn new() -> MVHashMap<K, V> {
+    pub fn new(code_cache: Option<VersionedCode<K, V, X>>) -> MVHashMap<K, V, X> {
         MVHashMap {
-            data: DashMap::new(),
-            delta_keys: Mutex::new(Vec::new()),
+            data: VersionedData::new(),
+            code: code_cache.unwrap_or_default(),
```
Comment (zekun000) [inline aptos-move/mvhashmap/src/lib.rs:46]: why do we need the arg code_cache?

File: aptos-move/mvhashmap/src/lib.rs [hunk 21]
```diff
@@ -2,304 +2,126 @@
...
     }
 
     /// Delete an entry from transaction 'txn_idx' at access path 'key'. Will panic
     /// if the access path has never been written before.
     pub fn delete(&self, key: &K, txn_idx: TxnIndex) {
-        // TODO: investigate logical deletion.
-        let mut v = self.data.get_mut(key).expect("Path must exist");
-        v.versioned_map.remove(&txn_idx);
+        match key.module_path() {
+            Some(_) => self.code.delete(key, txn_idx),
```
Comment (zekun000) [inline aptos-move/mvhashmap/src/lib.rs:67]: we probably should return error or assert on module deletion

File: aptos-move/mvhashmap/src/versioned_code.rs [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+        &self,
+        key: &K,
+        txn_idx: TxnIndex,
+    ) -> anyhow::Result<MVCodeOutput<V, X>, MVCodeError> {
+        use MVCodeError::*;
+        use MVCodeOutput::*;
+
+        match self.read(key, txn_idx) {
+            Ok((module, hash)) => {
+                let v = self.values.get(key).expect("Path must exist");
```
Comment (zekun000) [inline aptos-move/mvhashmap/src/versioned_code.rs:181]: this seems unnecessary if `read` returns the Value directly

File: types/src/executable.rs [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+pub trait ModulePath {
+    fn module_path(&self) -> Option<AccessPath>;
+}
+
+impl ModulePath for StateKey {
+    fn module_path(&self) -> Option<AccessPath> {
+        if let StateKeyInner::AccessPath(ap) = self.inner() {
+            if ap.is_code() {
```
Comment (zekun000) [inline types/src/executable.rs:27]: you probably are not aware that inside this function, it does a bcs deserialization each time, we should cache it on access path side if it's used heavily

**Review by zekun000 — APPROVED** [review]

File: aptos-move/block-executor/src/proptest_types/tests.rs (hunk 18 above)
Comment (gelash) [reply aptos-move/block-executor/src/proptest_types/tests.rs:409]: I think probably the hashing you pointed out for access_path (this one will go away) and also computing the crypto hash for every module - so it became slower as this one hammers module publishing path.

File: aptos-move/block-executor/src/scheduler.rs (hunk 19 above)
Comment (gelash) [reply aptos-move/block-executor/src/scheduler.rs:173]: great point - I think why it wasn't padded is since all threads were reading the validation and execution indices pretty often, and then having them actually padded in separate cache-lines could be even worse (2x invalidations). Never did precise benchmarks though - so it's all about how much false sharing penalty vs actual sharing overhead of having them separate :-) But (a) we can definitely pad commit_idx and if and when we have an optimization for validator and executor distinction to stick to threads, it will actually become important.

So maybe let's pad all those indices and if there is no regression keep that way, more forward looking. I'll add a comment to validation & execution indices though regarding the above.

File: aptos-move/mvhashmap/src/lib.rs (hunk 20 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:46]: to re-use between blocks.. I could delete and introduce later, probably will change the API anyway, so I'll get rid of it.

Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:46]: added a comment.

File: aptos-move/mvhashmap/src/lib.rs (hunk 21 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:67]: oh, just realized, this isn't module deletion, this is just for deleting an entry when execution aborts.

File: aptos-move/mvhashmap/src/versioned_code.rs (hunk 22 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_code.rs:181]: Moved read to be a method on VersionedValue, looks a lot cleaner now

File: types/src/executable.rs (hunk 23 above)
Comment (gelash) [reply types/src/executable.rs:27]: this will be static dispatch from the caller eventually when we have separate traits for code (move-vm knows). Will fix it if we see a regression.
Adding caching looked problematic because access_path entries are pub.
